pkg/controller/node/node_controller.go

```diff
@@ -976,22 +980,32 @@ func (ctrl *Controller) updateCandidateMachines(pool *mcfgv1.MachineConfigPool,
 	targetConfig := pool.Spec.Configuration.Name
 	for _, node := range candidates {
 
-		// Check image details for this pool
-		targetImage, equivalentTo, isImagePool, targetImageMatchesConfig := ctrl.experimentalHasValidImage(pool)
-
-		// If our pool is annotated with an image AND that image is the right image, then update these nodes with it
-		// TODO(jkyros): this does not obviate the desiredConfig annotation below for now, as we're using "are we in our desired machineconfig"
-		// as our "done" signal
-		if isImagePool {
-			if targetImageMatchesConfig {
-				ctrl.logPool(pool, "Setting node %s target image to %s", node.Name, targetImage)
-				if err := ctrl.setNodeAnnotation(node.Name, daemonconsts.DesiredImageConfigAnnotationKey, targetImage); err != nil {
-					return goerrs.Wrapf(err, "setting desired config for node %s", node.Name)
+		// double check that this node has CoreOS and will support layering
+		osID, ok := node.Labels[osIDLabel]
+		if ok {
+			if daemon.IsProbablyCoreOSVariant(osID) {
+				// Check image details for this pool
+				targetImage, equivalentTo, isImagePool, targetImageMatchesConfig := ctrl.experimentalHasValidImage(pool)
+
+				// If our pool is annotated with an image AND that image is the right image, then update these nodes with it
+				// TODO(jkyros): this does not obviate the desiredConfig annotation below for now, as we're using "are we in our desired machineconfig"
+				// as our "done" signal
+				if isImagePool {
+					if targetImageMatchesConfig {
+						ctrl.logPool(pool, "Setting node %s target image to %s", node.Name, targetImage)
+						if err := ctrl.setNodeAnnotation(node.Name, daemonconsts.DesiredImageConfigAnnotationKey, targetImage); err != nil {
+							return goerrs.Wrapf(err, "setting desired config for node %s", node.Name)
+						}
+					} else {
+						glog.Infof("Image %s matched %s not %s. Proper image may not have rendered yet", targetImage, equivalentTo, targetConfig)
+					}
+
 				}
 			} else {
-				glog.Infof("Image %s matched %s not %s. Proper image may not have rendered yet", targetImage, equivalentTo, targetConfig)
+				glog.Infof("falling back to legacy update for non-CoreOS node %s; node has OS ID %s", node.Name, osID)
```

Comment:
We've kind of been implicitly saying that "layering" is pool-granular, and it has to be once our content `spec.Configuration` is an Image. So while this works today with the annotations to where we can go either/or, this won't work anymore once we set `spec.Configuration` to an image, and I don't know that we want it to? 

If we don't like that it's pool-granular, we probably need to worry about those consequences now before we get too far. 

(If we're not okay with the trades and we do want "heterogeneous pools" then our pool probably has multiple payloads available that are vaguely equivalent and get assigned to nodes based on what kind they are, and we should probably mess with our CRD to accommodate that)

Reply:
I don't think we need to care about pool-granular right now *except* that we will likely need at some point to be able to handle in-place update transitions to images.